spec/abci++/abci++_properties_001_draft.md

```diff
@@ -0,0 +1,675 @@
+      * _p_'s Tendermint collects outstanding transactions from the mempool
+        * The transactions will be collected in order of priority
+        * Let $C$ the list of currently collected transactions
+        * The collection stops when any of the following conditions no longer holds
+          * the mempool is not empty
+          * the total size of transactions $\in C$ is less than `consensusParams.block.max_bytes`
+          * the sum of `GasWanted` field of transactions $\in C$ is less than `consensusParams.block.max_gas`
+      * _p_'s Tendermint creates a block header.
+      * _p_'s Tendermint uses the newly generated block as block _v_
+2. _p_'s Tendermint calls `RequestPrepareProposal` with block _v_.
+   The call is synchronous: Tendermint's execution will block until the Application returns from the call.
+3. The Application checks the block (header, transactions, commit info, evidences). Besides,
+   * in "same-block execution" mode, the Application can (and should) provide `ResponsePrepareProposal.data`,
+     `ResponsePrepareProposal.validator_updates`, or
+     `ResponsePrepareProposal.consensus_param_updates`.
+   * in "next-block execution" mode, _p_'s Tendermint will ignore the values for `ResponsePrepareProposal.data`,
+     `ResponsePrepareProposal.validator_updates`, and `ResponsePrepareProposal.consensus_param_updates`.
+   * in both modes, the Application can manipulate transactions
+      * leave transaction untouched
+      * add new transactions (not previously in the mempool)
+      * mark transactions is invalid (remove)
+      * remove transactions (effectively _delaying_ them)
+      * modify transactions (e.g. aggregate them)
+      * reorder transactions
+   * **TODO**: include logic of `VerifyVoteExtension` here?
+4. If the block is modified, the Application sets `ResponsePrepareProposal.modified` to true,
+   and includes the modified block in the return parameters (see the rules in section _Usage_).
+   The Application returns from the call.
+5. _p_'s Tendermint uses the (possibly) modified block as _p_'s proposal in round _r_, height _h_.
+
+### ProcessProposal
+
+#### Parameters and Types
+
+* **Request**:
+
+    | Name                 | Type                                        | Description                                                                                                 | Field Number |
+    |----------------------|---------------------------------------------|-------------------------------------------------------------------------------------------------------------|--------------|
+    | hash                 | bytes                                       | The proposed block's hash. This can be derived from the block header.                                       | 1            |
+    | header               | [Header](../core/data_structures.md#header) | The proposed block's header.                                                                                | 2            |
```

Comment:
I'm not sure we should reuse the same data structures here as those used in consensus. This creates a really tight coupling between our external API and the ABCI API.

Reply:
Ah, I see. I was more concerned about which types _at the code level_ we were going to use here. Having them contain the same data is totally fine, but I think having them be the same protos may be a bit risky but that is something we can discuss as we begin the implementation.